File: impls/rust/step1_read_print.rs
use std::io::{self, Write};

extern crate regex;

// This works ... but is clearly a wrongun.
mod types;
use types::MalVal;
mod reader;
use reader::read_str;
mod printer;
use printer::pr_str;

#[allow(non_snake_case)]
fn READ(code: String) -> MalVal {
    read_str(code)
}

#[allow(non_snake_case)]
fn EVAL(code: MalVal) -> MalVal {
    code
}

#[allow(non_snake_case)]
fn PRINT(code: MalVal) -> String {
    pr_str(code)
}

fn rep(code: String) -> String {
    PRINT(EVAL(READ(code)))
}

fn main() {
    loop {
        print!("user> ");
        io::stdout().flush().unwrap();

        let mut input = String::new();
        match io::stdin().read_line(&mut input) {
            Ok(_)  => { println!("{}", rep(input)); }
            Err(_) => { break; }
        }
    }
}

File: impls/ts/step1_read_print.ts
import { MalType } from './types.ts'
import { read_str } from './reader.ts'
import { pr_str } from './printer.ts'

function READ(s: string): MalType {
    return read_str(s)
}
function EVAL(v: MalType): MalType {
    return v
}
function PRINT(v: MalType): string {
    return pr_str(v)
}
function rep(s: string): string {
    return PRINT(EVAL(READ(s)))
}

let line = prompt('user>')
while (line !== null) {
    try {
        console.log(rep(line));
    }
    catch(e) {
        console.log('Error: ', e)
    }
    line = prompt('user>')
}